### DISCUSSION TITLE
Credentials in WmsProvider.CreateAsync

### DISCUSSION THREAD

**Discussion by Thoronion** [Mapsui/Mapsui#2466] [Q&A] [answered]
Hi. 

I am not able to figure out how to pass credentials to `WmsProvider.CreateAsync`. The `GetStreamAsync` in Client.cs is not handling credentials and throws an exception. My workaround is to implement my own `GetStreamAsync` and pass it to `WmsProvider.CreateAsync`.

What am I missing 😇

**Comment by pauldendulk** [comment]
There is WmsProvider.Credentials, did you try that?

```csharp
var provider = await WmsProvider.CreateAsync(wmsUrl);
provider.Credentials = new NetworkCredential("username", "password");
```

**Reply by Thoronion** [reply]
Thank you for your reply. 

That will not work.

The problem is that `WmsProvider.CreateAsync` is calling `Client.CreateAsync` which in turn makes an HTTP request to the specified WMS URL to get the capabilities. This happens before the provider is created and thus not using the provided credentials. This results in an unauthorized response which causes an exception from Mapsui. 

I am not sure if this is by design and I am missing something or if the credentials are really missing in Client.

**Reply by pauldendulk** [reply, marked as answer]
ah, I see. This is a common problem we have in our api design. The HttpClient is created deep inside. Also we do not want to offer all possible ways in which an HttpClient can be instantiated. Creating the HttpClient should be more on the surface, or on the outside.

In this case it looks like you can do you own request (in which you create your own HttpClient) by passing in the getStreamAsync function of Client.CreateAsync.
